src/input.c

```diff
@@ -48,20 +49,273 @@ FILE* check_args(int argc, char *argv[])
+  }
+  return line;
+}
+
+/*******************************************************************************************
+ * NAME :             read_clauses
+ *
+ * DESCRIPTION :      Progresses the file pointer past the commentes.
+ *
+ *
+ * INPUTS :
+ *      PARAMETERS :   
+ *          FILE*       fp                      The pointer to the file.
+ *          int         file size               The size of the file in bytes.
+ *          int**       data                    The pointer to the set of pointers to the clauses.
+ *          int*        clause_lengths          The array of clause lengths.
+ *          int*        unit_clauses            The array of clauses of 1 value.
+ *          int*        unit_clauses_length     The number of unit clauss.
+ *
+ * OUTPUTS :
+ *      RETURN :
+ *          int*                     1 on success, -1 on failure/error.
+ */
+int read_clauses(FILE* fp, int file_size, int** data, int* clause_lengths, int* unit_clauses, int* unit_clauses_length)
+{
+  int current_data_index = 0;
+  int current_clause_index = 0;
+  int clause_length = 0;
+
+  char* endptr; // For error checking strtol.
+
+  char* line = input_string(fp, file_size);
+
+  while(line != NULL && strcmp(line, ""))
+  {
+    // Create copy of line because strtok() consumes line during count.
+    char* line_copy = malloc(strlen(line));
+    strncpy(line_copy, line, strlen(line));
+
+    char* split_clause = strtok(line, " ");
+
+    // Count the number of values in a clause.
+    while(split_clause != NULL)
+    {
+      clause_length++;
+      split_clause = strtok (NULL, " "); // Increment to next value.
+    }
+
+    // Subtract 1 to account for clause terminating 0.
+    clause_length -= 1; 
+
+    // Sets clause length in clause_lengths array.
+    clause_lengths[current_data_index] = clause_length;
+
+    int* clause = (int *)malloc(sizeof(int) * clause_length); // A clause to hold values.//
+
+    // Start to load values. 
+    split_clause = strtok(line_copy, " ");
+
+    // Read clause values from char* and place into clause[].
+    while(split_clause != NULL)
+    {
+      // Tries to converts char* to int.
+      int clause_value = strtol(split_clause, &endptr, 0); 
+      clause[current_clause_index] = clause_value;
+
+      if(*endptr != '\0' && clause_value != 0) { return 0; }// Verify int value.
+      
+      split_clause = strtok (NULL, " "); // Increment to next value.
+      current_clause_index++; // Increment the clause index.
+    }
+
+    // If clause is only length 1, save value to molested intput struct.
+    if(clause_length == 1)
+    {
+      unit_clauses[*unit_clauses_length] = clause[current_clause_index-2];
+      (*unit_clauses_length)++;
+    }
+
+    data[current_data_index] = clause; // Set clause in data.
+
+    current_data_index++; // Increment the data index.
+    current_clause_index = 0; // Reset the clause index.
+    clause_length = 0;
+
+    line = input_string(fp, file_size);
+  }
+  return 1;
```

Comment:
It actually does. When endptr is being checked. And when I improve this function to have better input validation, it will likely have more return values.